// pareto-core-command/dist/actions.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.assert = exports.if_ = exports.refine_stacked = exports.refine_without_error_transformation = exports.query_stacked = exports.query_without_error_transformation = exports.sequence = exports.nothing = exports.create_error_handling_context = exports.dictionary = exports.list = exports.fail = void 0;
const _pinternals = __importStar(require("pareto-core-internals"));
const create_command_promise_1 = require("./create_command_promise");
const sequence_1 = require("./sequence");
const fail = (error) => {
    return (0, create_command_promise_1.__create_command_promise)({
        'execute': (on_success, on_error) => {
            on_error(error);
        }
    });
};
exports.fail = fail;
var list;
(function (list) {
    list.parallel = (the_array, errors_aggregator) => {
        return (0, create_command_promise_1.__create_command_promise)({
            'execute': (on_success, on_error) => {
                const errors_builder = _pinternals.create_asynchronous_list_builder();
                _pinternals.create_asynchronous_processes_monitor((monitor) => {
                    the_array.map(($) => {
                        monitor['report process started']();
                        $.__start(() => {
                            monitor['report process finished']();
                        }, (e) => {
                            errors_builder['add item'](e);
                            monitor['report process finished']();
                        });
                    });
                }, () => {
                    const errors = errors_builder['get list']();
                    if (errors.is_empty()) {
                        on_success();
                    }
                    else {
                        on_error(errors_aggregator(errors));
                    }
                });
            }
        });
    };
    list.serie = (array) => {
        return (0, create_command_promise_1.__create_command_promise)({
            'execute': (on_success, on_error) => {
                let current = 0;
                const do_next = () => {
                    array.__get_element_at(current).transform(($) => {
                        current += 1;
                        $.__start(() => {
                            do_next();
                        }, on_error);
                    }, () => {
                        on_success();
                    });
                };
                do_next();
            }
        });
    };
})(list || (exports.list = list = {}));
var dictionary;
(function (dictionary_1) {
    dictionary_1.parallel = (dictionary, block, aggregate_errors) => {
        return (0, create_command_promise_1.__create_command_promise)({
            'execute': (on_success, on_error) => {
                const errors_builder = _pinternals.create_asynchronous_dictionary_builder();
                _pinternals.create_asynchronous_processes_monitor((monitor) => {
                    dictionary.map(($, key) => {
                        monitor['report process started']();
                        (0, sequence_1.__sequence)(block($, key)).__start(() => {
                            monitor['report process finished']();
                        }, (e) => {
                            errors_builder['add entry'](key, e);
                            monitor['report process finished']();
                        });
                    });
                }, () => {
                    const errors = errors_builder['get dictionary']();
                    if (errors.is_empty()) {
                        on_success();
                    }
                    else {
                        on_error(aggregate_errors(errors));
                    }
                });
            }
        });
    };
    let deprecated_parallel;
    (function (deprecated_parallel) {
        deprecated_parallel.query = (staging_result, block, aggregate_errors) => {
            return (0, create_command_promise_1.__create_command_promise)({
                'execute': (on_success, on_error) => {
                    const errors_builder = _pinternals.create_asynchronous_dictionary_builder();
                    staging_result.__extract_data((dictionary) => {
                        _pinternals.create_asynchronous_processes_monitor((monitor) => {
                            dictionary.map(($, key) => {
                                monitor['report process started']();
                                (0, sequence_1.__sequence)(block($, key)).__start(() => {
                                    monitor['report process finished']();
                                }, (e) => {
                                    errors_builder['add entry'](key, e);
                                    monitor['report process finished']();
                                });
                            });
                        }, () => {
                            const errors = errors_builder['get dictionary']();
                            if (errors.is_empty()) {
                                on_success();
                            }
                            else {
                                on_error(aggregate_errors(errors));
                            }
                        });
                    }, on_error);
                }
            });
        };
    })(deprecated_parallel = dictionary_1.deprecated_parallel || (dictionary_1.deprecated_parallel = {}));
    // export const serie = <Error, Entry_Error>(
    //     dictionary: _pi.Dictionary<_pi.Command_Promise<Entry_Error>>,
    //     error_transformer: ($: Entry_Error, key: string) => Error,
    // ): _pi.Command_Promise<Error> => {
    //     return __create_command_promise({
    //         'execute': (on_success, on_error) => {
    //             const as_list = dictionary.to_list(($, key) => ({
    //                 'key': key,
    //                 'value': $,
    //             }))
    //             let current = 0
    //             const do_next = () => {
    //                 as_list.__get_element_at(current).transform(
    //                     ($) => {
    //                         const key = $.key
    //                         current += 1
    //                         $.value.__start(
    //                             () => {
    //                                 do_next()
    //                             },
    //                             ($) => {
    //                                 on_error(error_transformer($, key))
    //                             }
    //                         )
    //                     },
    //                     () => {
    //                         on_success()
    //                     }
    //                 )
    //             }
    //             do_next()
    //         }
    //     })
    // }
    // export const deprecated_parallel_without_transforming_the_error = <Error, Entry_Error>(
    //     dictionary: _pi.Dictionary<_pi.Command_Promise<Entry_Error>>,
    // ): _pi.Command_Promise<_pi.Dictionary<Entry_Error>> => {
    //     return __create_command_promise({
    //         'execute': (
    //             on_success,
    //             on_error,
    //         ) => {
    //             const errors_builder = _pinternals.create_asynchronous_dictionary_builder<Entry_Error>()
    //             _pinternals.create_asynchronous_processes_monitor(
    //                 (monitor) => {
    //                     dictionary.map(($, key) => {
    //                         monitor['report process started']()
    //                         $.__start(
    //                             () => {
    //                                 monitor['report process finished']()
    //                             },
    //                             (e) => {
    //                                 errors_builder['add entry'](key, e)
    //                                 monitor['report process finished']()
    //                             }
    //                         )
    //                     })
    //                 },
    //                 () => {
    //                     const errors = errors_builder['get dictionary']()
    //                     if (errors.is_empty()) {
    //                         on_success()
    //                     } else {
    //                         on_error(errors)
    //                     }
    //                 }
    //             )
    //         }
    //     })
    // }
})(dictionary || (exports.dictionary = dictionary = {}));
const create_error_handling_context = (command_block, processor, new_error) => {
    return (0, create_command_promise_1.__create_command_promise)({
        'execute': (on_success, on_error) => {
            (0, sequence_1.__sequence)(command_block).__start(on_success, (e) => {
                processor(e).__start(() => {
                    on_error(new_error);
                }, on_error);
            });
        }
    });
};
exports.create_error_handling_context = create_error_handling_context;
const nothing = () => {
    return (0, create_command_promise_1.__create_command_promise)({
        'execute': (on_success) => {
            on_success();
        }
    });
};
exports.nothing = nothing;
// Sequencing function:
const sequence = (block) => {
    return (0, create_command_promise_1.__create_command_promise)({
        'execute': (on_success, on_error) => {
            const length = _pinternals.list_literal(block).get_number_of_elements();
            const runStep = (index) => {
                if (index >= length) {
                    on_success();
                    return;
                }
                block[index].__start(() => runStep(index + 1), on_error);
            };
            runStep(0);
        }
    });
};
exports.sequence = sequence;
const query_without_error_transformation = (staging_result, command_block) => {
    return (0, create_command_promise_1.__create_command_promise)({
        'execute': (on_success, on_error) => {
            staging_result.__extract_data((output) => {
                (0, sequence_1.__sequence)(command_block(output)).__start(() => {
                    on_success();
                }, (e) => {
                    on_error(e);
                });
            }, (e) => {
                on_error(e);
            });
        }
    });
};
exports.query_without_error_transformation = query_without_error_transformation;
const query_stacked = (staging_result, parent_data, command_block) => {
    return (0, create_command_promise_1.__create_command_promise)({
        'execute': (on_success, on_error) => {
            staging_result.__extract_data((output) => {
                (0, sequence_1.__sequence)(command_block(output, parent_data)).__start(() => {
                    on_success();
                }, (e) => {
                    on_error(e);
                });
            }, on_error);
        }
    });
};
exports.query_stacked = query_stacked;
const refine_without_error_transformation = (callback, command_block) => {
    return (0, create_command_promise_1.__create_command_promise)({
        'execute': (on_success, on_error) => {
            _pinternals.deprecated_create_refinement_context(callback).__extract_data((output) => {
                (0, sequence_1.__sequence)(command_block(output)).__start(on_success, on_error);
            }, (e) => {
                on_error(e);
            });
        }
    });
};
exports.refine_without_error_transformation = refine_without_error_transformation;
const refine_stacked = (callback, parent_data, command_block) => {
    return (0, create_command_promise_1.__create_command_promise)({
        'execute': (on_success, on_error) => {
            _pinternals.deprecated_create_refinement_context(callback).__extract_data((output) => {
                (0, sequence_1.__sequence)(command_block(output, parent_data)).__start(() => {
                    on_success();
                }, (e) => {
                    on_error(e);
                });
            }, on_error);
        }
    });
};
exports.refine_stacked = refine_stacked;
var if_;
(function (if_) {
    if_.direct = (precondition, command_block, else_command_block) => {
        return (0, create_command_promise_1.__create_command_promise)({
            'execute': (on_success, on_error) => {
                if (precondition) {
                    (0, sequence_1.__sequence)(command_block).__start(on_success, on_error);
                }
                else {
                    if (else_command_block !== undefined) {
                        (0, sequence_1.__sequence)(else_command_block).__start(on_success, on_error);
                    }
                    else {
                        on_success();
                    }
                }
            }
        });
    };
    if_.query = (precondition, command_block, else_command_block) => {
        return (0, create_command_promise_1.__create_command_promise)({
            'execute': (on_success, on_error) => {
                precondition.__extract_data(($) => {
                    if ($) {
                        (0, sequence_1.__sequence)(command_block).__start(on_success, on_error);
                    }
                    else {
                        if (else_command_block !== undefined) {
                            (0, sequence_1.__sequence)(else_command_block).__start(on_success, on_error);
                        }
                        else {
                            on_success();
                        }
                    }
                }, on_error);
            }
        });
    };
})(if_ || (exports.if_ = if_ = {}));
var assert;
(function (assert) {
    assert.direct = (assertion, error_if_failed) => {
        return (0, create_command_promise_1.__create_command_promise)({
            'execute': (on_success, on_error) => {
                if (!assertion) {
                    on_error(error_if_failed);
                    return;
                }
                on_success();
            }
        });
    };
    assert.query = (assertion, error_if_failed) => {
        return (0, create_command_promise_1.__create_command_promise)({
            'execute': (on_success, on_error) => {
                assertion.__extract_data(($) => {
                    if ($) {
                        on_success();
                    }
                    else {
                        on_error(error_if_failed);
                    }
                }, ($) => {
                    on_error($);
                });
            }
        });
    };
})(assert || (exports.assert = assert = {}));
