// pareto-core-query/dist/create_query_result.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.__create_query_result = __create_query_result;
const _pinternals = __importStar(require("pareto-core-internals"));
class Query_Result_Class {
    constructor(executer) {
        this.executer = executer;
        this.query_result = null;
    }
    transform_result(transformer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                on_result(transformer($));
            }, on_error);
        });
    }
    deprecated_transform_error(error_transformer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(on_result, ($) => {
                on_error(error_transformer($));
            });
        });
    }
    query_without_error_transformation(queryer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                queryer($).__extract_data(on_result, on_error);
            }, on_error);
        });
    }
    query(queryer, error_transformer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                queryer($).__extract_data(on_result, (queryer_error) => {
                    on_error(error_transformer(queryer_error));
                });
            }, on_error);
        });
    }
    refine_without_error_transformation(callback) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                _pinternals.deprecated_create_refinement_context((abort) => callback($, abort)).__extract_data(on_result, on_error);
            }, on_error);
        });
    }
    refine(callback, error_transformer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(($) => {
                _pinternals.deprecated_create_refinement_context((abort) => callback($, abort)).__extract_data(on_result, (stager_error) => {
                    on_error(error_transformer(stager_error));
                });
            }, on_error);
        });
    }
    rework_error_temp(error_reworker, rework_error_transformer) {
        return new Query_Result_Class((on_result, on_error) => {
            this.executer(on_result, ($) => {
                error_reworker($).__extract_data((new_target_error) => {
                    on_error(new_target_error);
                }, (rework_error) => {
                    on_error(rework_error_transformer(rework_error));
                });
            });
        });
    }
    __extract_data(on_result, on_error) {
        this.executer(on_result, on_error);
    }
}
function __create_query_result(executer) {
    return new Query_Result_Class(executer);
}
