// pareto-core-refiner/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deprecated_build_dictionary = exports.ss = exports.au = exports.cc = exports.list_to_dictionary = exports.fixme_abort = exports.deprecated_cc = exports.set = exports.not_set = exports.list_literal = exports.dictionary_literal = exports.create_iterator = exports.build_dictionary = exports.build_list = exports.block = void 0;
const _pinternals = __importStar(require("pareto-core-internals"));
/**
 * these functions coming from core-internals should be exposed for deserializer development
 */
var pareto_core_internals_1 = require("pareto-core-internals");
Object.defineProperty(exports, "block", { enumerable: true, get: function () { return pareto_core_internals_1.block; } });
Object.defineProperty(exports, "build_list", { enumerable: true, get: function () { return pareto_core_internals_1.build_list; } });
Object.defineProperty(exports, "build_dictionary", { enumerable: true, get: function () { return pareto_core_internals_1.build_dictionary; } });
Object.defineProperty(exports, "create_iterator", { enumerable: true, get: function () { return pareto_core_internals_1.create_iterator; } });
Object.defineProperty(exports, "dictionary_literal", { enumerable: true, get: function () { return pareto_core_internals_1.dictionary_literal; } });
Object.defineProperty(exports, "list_literal", { enumerable: true, get: function () { return pareto_core_internals_1.list_literal; } });
Object.defineProperty(exports, "not_set", { enumerable: true, get: function () { return pareto_core_internals_1.not_set; } });
Object.defineProperty(exports, "set", { enumerable: true, get: function () { return pareto_core_internals_1.set; } });
Object.defineProperty(exports, "deprecated_cc", { enumerable: true, get: function () { return pareto_core_internals_1.cc; } });
Object.defineProperty(exports, "fixme_abort", { enumerable: true, get: function () { return pareto_core_internals_1.panic; } });
Object.defineProperty(exports, "list_to_dictionary", { enumerable: true, get: function () { return pareto_core_internals_1.list_to_dictionary; } });
Object.defineProperty(exports, "cc", { enumerable: true, get: function () { return pareto_core_internals_1.cc; } });
Object.defineProperty(exports, "au", { enumerable: true, get: function () { return pareto_core_internals_1.au; } });
Object.defineProperty(exports, "ss", { enumerable: true, get: function () { return pareto_core_internals_1.ss; } });
const deprecated_build_dictionary = ($) => {
    const temp = {};
    $({
        'add entry': (key, $) => {
            if (key in temp) {
                _pinternals.panic(`duplicate key in dictionary literal: ${key}`);
            }
            temp[key] = $;
        }
    });
    return _pinternals.dictionary_literal(temp);
};
exports.deprecated_build_dictionary = deprecated_build_dictionary;
